flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/IntervalJoinITCase.java

```diff
@@ -155,6 +157,134 @@ public void testJoinsCorrectlyWithMultipleKeys() throws Exception {
                 "(key2,5):(key2,5)");
     }
 
+    @Test
+    public void testIntervalJoinSideOutputLeftLateData() throws Exception {
```

Comment:
The 2 tests are 99% the same. Can we do something to merge then into a single function call w/ parameters?
For example the content of `run` can be a function pointer to the extracted function or something.

Reply:
Thanks for your tips! I have merged the 2 tests into one test method by sideout the  late data of left and right side at the same time .

Should I continue to  extracted  the `run` content ? Cause I am a little confuse about how to do that.

Reply:
The technique is create a function like
```
private void testIntervalJoinSideOutput(Consumer<SourceContext<Tuple2<String, Integer>>> streamOneRun) {
...
        DataStream<Tuple2<String, Integer>> streamOne =
                env.addSource(
                        new SourceFunction<Tuple2<String, Integer>>() {
                            @Override
                            public void run(SourceContext<Tuple2<String, Integer>> ctx) {
                                streamOneRun.accept(ctx);
                            }

                            @Override
                            public void cancel() {
                                // do nothing
                            }
                        });
...
}
```

and call it like this:
```
    @Test
    public void testIntervalJoinSideOutputLeftLateData() throws Exception {
        testIntervalJoinSideOutput(ctx -> {
          ctx.collectWithTimestamp(Tuple2.of("key", 2), 2L);
          ctx.collectWithTimestamp(Tuple2.of("key", 3), 3L);
          ctx.emitWatermark(new Watermark(3));
          ctx.collectWithTimestamp(Tuple2.of("key", 1), 1L); // late data
        });
...
        expectInAnyOrder("(key,1)");
}
```
Be aware that this may or may not compile, just wanted to show the way :)

Reply:
Well, I understand what you mean. I have extract the `run` content as a `Consumer` func